### PR TITLE
Windows 10/11 Calculator: report correct number after paste by fetching the value directly from the UI element.

### LINKED ISSUE #16573: Windows 10/11 Calculator: NVDA does not announce the entire expression when pasting it from the clipboard
Hi,

Discussed on the following NVDA users list thread: https://nvda.groups.io/g/nvda/topic/nvda_2024_2_beta_1_and/106097471

### Steps to reproduce:
Prerequisite: Windows 10 or 11:

1. Open Notepad or other text editors or move to an edit field.
2. Type a number or an expression (say, 12.34).
3. Copy the number to the clipboard.
4. Open Calculator.
5. Press Control+V to paste the copied nubmer into the display field.

### Actual behavior:
NVDA does not announce the pasted number/expression completely (for example, NVDA says "12" when pasting 12.34).

### Expected behavior:
NVDA announces the pasted number completely.

### NVDA logs, crash dumps and other attachments:
None

#### NVDA installed/portable/running from source:
Installed

#### NVDA version:
alpha-31971,c5954630 x86

#### Windows version:
Windows 11 23H2 beta (22635.3640)

#### Name and version of other software in use when reproducing the issue:
Windows Calculator 11.2403.6.0

#### Other information about your system:
Used as a development workstation

#### Does the issue still occur after restarting your computer?
Yes

#### Have you tried any other versions of NVDA? If so, please report their behaviors.
Reproducible with 2024.2 (beta) and earlier

#### If NVDA add-ons are disabled, is your problem still occurring?
Yes

#### Does the issue still occur after you run the COM Registration Fixing Tool in NVDA's tools menu?
N/A

### Technical information:
When the display field (housing expressions) change, both UIA notification and property/name change events are fired. At the moment NVDA ignores name change event except in specific cases; name change event coming from display field is ignored. When a number or an expression is pasted from the clipboard, NVDA will report whatever the UIA notification says, and unfortunately, the display string reported by this event is incomplete (more so if numbers after the decimal point are present). Unlike notification event, name change event does report the entire number/expression, but NVDA will catch it if the focus is on the display field (effect of selective UIA event registration).

There are at least two possible solutions, none of which will resolve this issue completely:

1. Allow limited handling of name change event in the display (calculator results) field. This works if focus is on the display field itself.
2. In UIA notification event handler, check both the display string and the field name/value for the element raising the notification. This doesn't work as the name of the element firing this event is set to an empty string.

This requires either a fix from Microsoft to refine Calculator's UIA notification event, or revisiting the decision to set event registration to selective in Windows 11 22H2 and later and subsequent issues identified. If we have no intention to revisit selective event registration decision, then the next best thing would be alerting Microsoft folks about UIA notification display string issue (if Microsoft does say notification event does present the entire number (say, "12.34" instead of "12"), then the ball comes back to our side of the court.

Thanks.

### PR DESCRIPTION
**Pull request by tburc**
### Link to issue number:

Fixes https://github.com/nvaccess/nvda/issues/16573

### Summary of the issue:

When pasting into the calculator, both speech and braille reported a truncated result (for example 1 instead of 12.34)

### Description of user facing changes

When pasting into the Windows 10/11 calculator, the correct number is announced.

### Description of development approach

* While handling UIA notification events, NVDA may announce screen changes not caused by number entry or results commands, and this is what is being handled (pasting from clipboard, for example).
* NVDA will traverse through result element's children, looking for the actual result.
* NVDA will announce the element's content once actual result element is identified.

### Testing strategy:

Manual testing. Specifically, test to make sure that NVDA is announcing the whole number when pasted (example: 12.34 instead of just 1). Also, make sure that NVDA only announces said number once.

### Known issues with pull request:

Nothing I can think of.

### Code Review Checklist:

- [x] Documentation: 
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech 
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [x] Security precautions taken.

## Summary by CodeRabbit

- **New Features**
  - Improved NVDA screen reader support for Windows 10/11 Calculator. NVDA now correctly reports the full number pasted into the calculator.

### CODE REVIEW

**Comment by josephsl** [comment]
Hello Tristan,

Thank you for your first contribution. As the original author of Calculator app module, I appreciate the effort taken to resolve this issue and update the code logic (and thank you for pointing out how tricky the logic is).

A few general comments:

* I love the honesty when you put "trial and error" in development strategy (yep, even experienced project members write that, so you are not alone). In the future, please provide at least a high-level overview of the proposed code changes (for an example, see #16555)..
* Calculator app module contents came from an NVDA add-on called Windows App Essentials (https://github.com/josephsl/wintenapps). When you read the commit log for that project, you will see some emotional content, including frustrations with app accessibility and usability.
* Speaking of code and its origins, the tricky logic part stems from "debates" between Microsoft, apps, yours truly, and NVDA - whenever Microsoft made changes to their apps, I would test them and offer feedback on accessibility for screen reader users, explaining how the code is organized and not so organized at the same time (my apologies); this is a side effect of Windows as a Service (WaaS) era - continuous development, continuous refinements, ongoing hunt for accessibility and usability issues.

Enough musings - if I understand the proposed code changes correctly:

1. While handling UIA notification events, NVDA may announce screen changes not caused by numer entry or results commands, and this is what is being handled (pasting from clipboard, for example).
2. NVDA will traverse through result element's children, looking for actual result text (using a for loop is smart).
3. NVDA will announce element content once actual result element is identified.

I don't expect new folks to be as detailed as above. I'll provide specific comments in the code fragments.

Thanks.

File: source/appModules/calculator.py [hunk 1]
```diff
@@ -6,13 +6,30 @@
 """App module for Windows 10 Calculator.
 This is also the base app module for Windows 11 Calculator."""
 
+# This file has some fairly tricky logic, so here is a test plan to make sure
```
Comment (josephsl) [inline source/appModules/calculator.py:9]: You don't have to write a test plan in the app module files - I think we should open a new discussion on adding test plans for Microsoft Store apps such as Calculator.

File: source/appModules/calculator.py [hunk 2]
```diff
@@ -112,6 +128,16 @@ def event_UIA_notification(self, obj, nextHandler, displayString=None, activityI
 				and doNotAnnounceCalculatorResults
 			):
 				return
+			# #16573: when pasting, calculator sends a truncated UIA displayString
+			# (for example "Display is 1" when it should be "Display is 12.34").
+			# To fix this, we ignore the displayString sent via the UIA notification,
+			# and fetch the value directly from the UI element.
+			for child in resultElement.children:
```
Comment (josephsl) [inline source/appModules/calculator.py:135]: Good catch - in full screen, expression text is marked as first child, whereas  in compact (mini) mode, actual results text is the first child.

File: source/appModules/calculator.py [hunk 3]
```diff
@@ -112,6 +128,16 @@ def event_UIA_notification(self, obj, nextHandler, displayString=None, activityI
...
+				if child.UIAAutomationId in ("CalculatorResults", "CalculatorAlwaysOnTopResults"):
+					speech.cancelSpeech()
+					speech.speakMessage(child.name)
```
Comment (josephsl) [inline source/appModules/calculator.py:138]:
```suggestion
					ui.message(child.name)
```

Comment (josephsl) [reply source/appModules/calculator.py:138]: Also applies to braille handler call. The ui.message function will both speak and braille whatever you want NVDA to say, so you can replace the speech and braille call with ui.message function call.

File: source/appModules/calculator.py [hunk 4]
```diff
@@ -112,6 +128,16 @@ def event_UIA_notification(self, obj, nextHandler, displayString=None, activityI
 				and doNotAnnounceCalculatorResults
 			):
 				return
+			# #16573: when pasting, calculator sends a truncated UIA displayString
+			# (for example "Display is 1" when it should be "Display is 12.34").
+			# To fix this, we ignore the displayString sent via the UIA notification,
+			# and fetch the value directly from the UI element.
+			for child in resultElement.children:
+				if child.UIAAutomationId in ("CalculatorResults", "CalculatorAlwaysOnTopResults"):
```
Comment (josephsl) [inline source/appModules/calculator.py:136]: What would happen if the display string reported by UIA notification is the same as what's shown on screen?

**Review by josephsl — CHANGES_REQUESTED** [review]
Overall, this looks good. I advise reading NVDA development guide (or additional parts of the source code) so you can learn a few tricks here and there (one of them is ui.message which allows NVDA to both speak and braille whatever). If you have time, please do read Git commit logs for Windows App Essentials add-on so you can see how the Calculator ap module came to be.

**Comment by seanbudd** [comment]
Hello Tristan, please use a descriptive title that explains what the PR does standalone.

File: source/appModules/calculator.py (hunk 1 above)
Comment (TristanBurchett) [reply source/appModules/calculator.py]: Ok, I deleted the test plan for now.  If the team does eventually decide you'd like to have such a thing, I'm happy to re-add it wherever you think it should go.

File: source/appModules/calculator.py (hunk 2 above)
Comment (TristanBurchett) [reply source/appModules/calculator.py:119]: Thanks!

File: source/appModules/calculator.py (hunk 3 above)
Comment (TristanBurchett) [reply source/appModules/calculator.py]: Great suggestion.  Done.

File: source/appModules/calculator.py (hunk 4 above)
Comment (TristanBurchett) [reply source/appModules/calculator.py:120]: I'm not sure I understand that situation you're considering, but I think the fix would still be applicable to that situation. For example, if you type "1+2=", the calculator will still show the correct answer and NVDA will report it correctly as well.  The UIA notification in this scenario would set displayText="Display is 3", and that string is exactly the same as what we would pull from the element itself.  So effectively, if the previous UIA notification was correct, then this does not change any behavior.

**Comment by TristanBurchett** [comment]
@josephsl thanks so much for your insightful and kind comments.  The code is quite good and I particularly appreciated how well commented it is.  It must have taken a lot of work to get it to this point.  This is my first time working with the NVDA project so I'm just not used to all the patterns involved.

**Comment by josephsl** [comment]
Hello Tristan,

At least we got the basics of the fix right, and thank you for your willingness to think about our feedback/comments.

I'm known for writing lots of comments in my own code - maybe I can't "keep my mouth shut" while programming... I think programming is a variation on writing essays (more so if you think about Python syntax and grammar). After all, working on apps like Calculator gave me writing materials in graduate school (I am attending a university in western United States).

About the for loop logic: consider the following scenario:

1. User presses Escape to clear Calculator display string.
2. A UIA notification event is fired.
3. NVDA notices that it can announce the display string since the last entered command is not a number.
4. The just proposed code (for loop) runs, announcing "display is 0".
5. Somehow notification event is handled once again, either becuase Calculator fired it or NVDA itself is handling it (again).
6. The display string announcmenet may repeat.

The key is checking for value equality (== versus !=); a possible optimization is that NVDA doesn't have to fall back to default UIA notification event handling implementation (where it will report whatever the display string is as long as the element is focused on screen). This is what the call to "nextHandler" is - NVDA's event handling system could be viewed as an iterative "chain" where different modules can observe and handle events if they want to, otherwise default behavior (such as announcing things) will be performed. I think we can safely ignore this odd thing in this PR.

If you want this to be part of NVDA 2024.3, I advise adding a changelog entry (we use Markdown syntax); if you want it in the next release after that, wait until 2024.3 enters beta phase (or shortly before) once NV Access people updates NVDA version in the source code. Then mark this PR as ready for review. Of course, since the issue may as well lie with Calculator executable itself, sending a "love letter" a.k.a. feedback to Microsoft from time to time about UIA notification event issue would be nice too.

Thanks.

File: source/appModules/calculator.py (hunk 1 above)
Comment (seanbudd) [reply source/appModules/calculator.py]: Please add manual test plans to `tests/manual`. Manual test plans are welcome per the contributing guide https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist

File: source/appModules/calculator.py [hunk 5]
```diff
@@ -112,6 +112,17 @@ def event_UIA_notification(self, obj, nextHandler, displayString=None, activityI
...
+					# When pasting, we get two UIA notification events.  Cancel the first
+					# one once we receive the second.
```
Comment (seanbudd) [inline source/appModules/calculator.py:122]:
```suggestion
					# When pasting, we get two UIA notification events. Cancel the first
					# one once we receive the second.
```

**Review by seanbudd — CHANGES_REQUESTED** [review]
Please add an entry to the changelog.
The general approach here looks good to me for the code changes.

**Comment by TristanBurchett** [comment]
> Please add an entry to the changelog. The general approach here looks good to me for the code changes.

Done.

**Comment by TristanBurchett** [comment]
> About the for loop logic: consider the following scenario:
> 
> 1. User presses Escape to clear Calculator display string.
> 2. A UIA notification event is fired.
> 3. NVDA notices that it can announce the display string since the last entered command is not a number.
> 4. The just proposed code (for loop) runs, announcing "display is 0".
> 5. Somehow notification event is handled once again, either becuase Calculator fired it or NVDA itself is handling it (again).
> 6. The display string announcement may repeat.

Using speech.cancelSpeech() eliminates the issue of duplicated speech announcements from UIA notifications.  In fact, when pasting into the calculator, this is exactly what happens -- you get two notifications.  So that's why I added the cancelSpeech() call.

**Review by seanbudd — APPROVED** [review]

**Pull request merged**